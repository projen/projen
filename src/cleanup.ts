import * as fs from "fs";
import * as path from "path";
import * as glob from "fast-glob";
import { PROJEN_DIR, PROJEN_MARKER } from "./common";
import * as logging from "./logging";

export const FILE_MANIFEST = `${PROJEN_DIR}/files.json`;

export function cleanup(dir: string, newFiles: string[], exclude: string[]) {
  try {
    const manifestFiles = getFilesFromManifest(dir);
    if (manifestFiles.length > 0) {
      // Use `FILE_MANIFEST` to remove files that are no longer managed by projen
      removeFiles(findOrphanedFiles(dir, manifestFiles, newFiles));
    } else {
      // Remove all files managed by projen with legacy logic
      removeFiles(findGeneratedFiles(dir, exclude));
    }
  } catch (e: any) {
    logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
  }
}

function removeFiles(files: string[]) {
  for (const file of files) {
    fs.rmSync(file, { force: true, recursive: true });
  }
}

function findGeneratedFiles(dir: string, exclude: string[]) {
  const ignore = [
    ...readGitIgnore(dir),
    "node_modules/**",
    ...exclude,
    ".git/**",
  ];

  const files = glob.sync("**", {
    ignore,
    cwd: dir,
    dot: true,
    onlyFiles: true,
    followSymbolicLinks: false,
    absolute: true,
  });

  const generated = new Array<string>();

  for (const file of files) {
    const contents = fs.readFileSync(file, "utf-8");

    if (contents.includes(PROJEN_MARKER)) {
      generated.push(file);
    }
  }

  return generated;
}

function findOrphanedFiles(
  dir: string,
  oldFiles: string[],
  newFiles: string[]
) {
  return oldFiles
    .filter((old) => !newFiles.includes(old))
    .map((f: string) => path.resolve(dir, f));
}

function getFilesFromManifest(dir: string): string[] {
  try {
    const fileManifestPath = path.resolve(dir, FILE_MANIFEST);
    if (fs.existsSync(fileManifestPath)) {
      const fileManifest = JSON.parse(
        fs.readFileSync(fileManifestPath, "utf-8")
      );
      if (fileManifest.files) {
        return fileManifest.files;
      }
    }
  } catch (e: any) {
    logging.warn(
      `warning: unable to get files to clean from file manifest: ${e.stack}`
    );
  }

  return [];
}

function readGitIgnore(dir: string) {
  const filepath = path.join(dir, ".gitignore");
  if (!fs.existsSync(filepath)) {
    return [];
  }

  return fs
    .readFileSync(filepath, "utf-8")
    .split("\n")
    .filter((x) => x?.trim() !== "")
    .filter((x) => !x.startsWith("#") && !x.startsWith("!"))
    .map((x) => x.replace(/^\//, "")) // remove "/" prefix
    .map((x) => `${x}\n${x}/**`)
    .join("\n")
    .split("\n");
}
