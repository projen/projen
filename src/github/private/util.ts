export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
