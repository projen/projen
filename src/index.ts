export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
