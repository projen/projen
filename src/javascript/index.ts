export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
