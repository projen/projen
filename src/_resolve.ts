import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
