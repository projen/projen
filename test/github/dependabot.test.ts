import {
  Dependabot,
  DependabotGroupAppliesTo,
  DependabotGroupDependencyType,
  DependabotGroupUpdateType,
  DependabotRegistryType,
} from "../../src/github";
import { NodeProject, NodeProjectOptions } from "../../src/javascript";
import { synthSnapshot } from "../util";

describe("dependabot", () => {
  test("default", () => {
    const project = createProject();

    new Dependabot(project.github!);

    const snapshot = synthSnapshot(project);
    expect(snapshot[".github/dependabot.yml"]).toBeDefined();
    expect(snapshot[".github/dependabot.yml"]).toMatchSnapshot();
  });

  test("private registries", () => {
    const project = createProject();

    const registryName = "npm-registry-npm-pkg-github-com";

    new Dependabot(project.github!, {
      registries: {
        [registryName]: {
          type: DependabotRegistryType.NPM_REGISTRY,
          url: "https://npm.pkg.github.com",
          token: "${{ secrets.TOKEN }}",
          replacesBase: true,
          username: "test",
          password: "${{ secrets.TOKEN }}",
          key: "abc123",
        },
      },
    });

    const snapshot = synthSnapshot(project);
    const dependabot = snapshot[".github/dependabot.yml"];
    expect(dependabot).toBeDefined();
    expect(dependabot).toMatchSnapshot();
    expect(dependabot).toContain("registries");
    expect(dependabot).toContain(registryName);
  });

  describe("groups", () => {
    test("no groups", () => {
      const project = createProject();
      new Dependabot(project.github!, {});
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("groups:");
      expect(dependabot).not.toContain("patterns:");
      expect(dependabot).not.toContain("exclude-patterns:");
    });

    test("two groups", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        groups: {
          groupOne: {
            patterns: ["testlib-*"],
          },
          groupTwo: {
            patterns: ["*"],
            excludePatterns: ["otherlib-*"],
          },
        },
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("groups:");
      expect(dependabot).toContain("patterns:");
      expect(dependabot).toContain("exclude-patterns:");
    });

    test("applies-to", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        groups: {
          groupOne: {
            appliesTo: DependabotGroupAppliesTo.SECURITY_UPDATES,
            patterns: ["*"],
          },
        },
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("groups:");
      expect(dependabot).toContain("applies-to: security-updates");
    });

    test("dependency-type", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        groups: {
          groupOne: {
            dependencyType: DependabotGroupDependencyType.DEVELOPMENT,
            patterns: ["*"],
          },
        },
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("groups:");
      expect(dependabot).toContain("dependency-type: development");
    });

    test("update-types", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        groups: {
          groupOne: {
            updateTypes: [
              DependabotGroupUpdateType.MINOR,
              DependabotGroupUpdateType.PATCH,
            ],
            patterns: ["*"],
          },
        },
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("groups:");
      expect(dependabot).toContain("update-types:");
      expect(dependabot).toContain("- minor");
      expect(dependabot).toContain("- patch");
    });

    test("explicitly empty update-types", () => {
      const project = createProject();
      expect(() => {
        new Dependabot(project.github!, {
          groups: {
            groupOne: {
              updateTypes: [],
              patterns: ["*"],
            },
          },
        });
      }).toThrow("empty");
    });

    test("duplicate update-types", () => {
      const project = createProject();
      expect(() => {
        new Dependabot(project.github!, {
          groups: {
            groupOne: {
              updateTypes: [
                DependabotGroupUpdateType.MAJOR,
                DependabotGroupUpdateType.MAJOR,
              ],
              patterns: ["*"],
            },
          },
        });
      }).toThrow("duplicate");
    });
  });

  describe("allowing", () => {
    test("allows testlib only", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        allow: [{ dependencyName: "testlib" }],
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("allow:");
      expect(dependabot).toContain("dependency-name: testlib");
    });
  });

  describe("ignoring", () => {
    test("ignores projen by default", () => {
      const project = createProject();
      new Dependabot(project.github!, {});
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("ignore:");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("ignore with ignoreProjen set to false", () => {
      const project = createProject();
      new Dependabot(project.github!, { ignoreProjen: false });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("ignore:");
      expect(dependabot).not.toContain("dependency-name: projen");
    });

    test("ignore with no version", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        ignore: [{ dependencyName: "testlib" }],
      });

      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("ignore");
      expect(dependabot).toContain("dependency-name: testlib");
      expect(dependabot).not.toContain("versions");
    });

    test("ignore with a single version", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        ignore: [{ dependencyName: "testlib", versions: [">10.x"] }],
      });

      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("ignore");
      expect(dependabot).toContain("dependency-name: testlib");
      expect(dependabot).toContain("versions");
      expect(dependabot).toContain(">10.x");
    });

    test("ignore with multiple versions", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        ignore: [{ dependencyName: "testlib", versions: ["10.x", "20.x"] }],
      });

      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("ignore");
      expect(dependabot).toContain("dependency-name: testlib");
      expect(dependabot).toContain("versions");
      expect(dependabot).toContain("10.x");
      expect(dependabot).toContain("20.x");
    });
  });

  describe("open-pull-requests-limit", () => {
    test("open pull requests limit set to positive integer", () => {
      const project = createProject();
      new Dependabot(project.github!, { openPullRequestsLimit: 5 });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("open-pull-requests-limit: 5");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("open pull requests limit set to 0", () => {
      const project = createProject();
      new Dependabot(project.github!, { openPullRequestsLimit: 0 });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("open-pull-requests-limit: 0");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("open pull requests limit not set", () => {
      const project = createProject();
      new Dependabot(project.github!);
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("open-pull-requests-limit:");
      expect(dependabot).toContain("dependency-name: projen");
    });
  });

  describe("assignees", () => {
    test("one assignee", () => {
      const project = createProject();
      new Dependabot(project.github!, { assignees: ["testUserOne"] });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("assignees:");
      expect(dependabot).toContain("testUserOne");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("multiple assignees", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        assignees: ["testUserOne", "testUserTwo"],
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("assignees:");
      expect(dependabot).toContain("testUserOne");
      expect(dependabot).toContain("testUserTwo");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("empty assignees", () => {
      const project = createProject();
      new Dependabot(project.github!, { assignees: [] });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("assignees:");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("no assignees", () => {
      const project = createProject();
      new Dependabot(project.github!);
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("assignees:");
      expect(dependabot).toContain("dependency-name: projen");
    });
  });

  describe("reviewers", () => {
    test("one reviewer", () => {
      const project = createProject();
      new Dependabot(project.github!, { reviewers: ["testUserOne"] });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("reviewers:");
      expect(dependabot).toContain("testUserOne");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("multiple reviewers", () => {
      const project = createProject();
      new Dependabot(project.github!, {
        reviewers: ["testUserOne", "testUserTwo"],
      });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain("reviewers:");
      expect(dependabot).toContain("testUserOne");
      expect(dependabot).toContain("testUserTwo");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("empty reviewers", () => {
      const project = createProject();
      new Dependabot(project.github!, { reviewers: [] });
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("reviewers:");
      expect(dependabot).toContain("dependency-name: projen");
    });

    test("no reviewers", () => {
      const project = createProject();
      new Dependabot(project.github!);
      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("reviewers:");
      expect(dependabot).toContain("dependency-name: projen");
    });
  });

  describe("target branch", () => {
    test("empty target branch", () => {
      const project = createProject();

      new Dependabot(project.github!, {});

      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toBeDefined();
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).not.toContain("target-branch4");
    });

    test("develop as the target branch", () => {
      const project = createProject();

      const targetBranch = "develop";

      new Dependabot(project.github!, {
        targetBranch: targetBranch,
      });

      const snapshot = synthSnapshot(project);
      const dependabot = snapshot[".github/dependabot.yml"];
      expect(dependabot).toBeDefined();
      expect(dependabot).toMatchSnapshot();
      expect(dependabot).toContain(`target-branch: ${targetBranch}`);
    });
  });
});

type ProjectOptions = Omit<
  NodeProjectOptions,
  "outdir" | "defaultReleaseBranch" | "name"
>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    defaultReleaseBranch: "main",
    name: "node-project",
    ...options,
  });
}
