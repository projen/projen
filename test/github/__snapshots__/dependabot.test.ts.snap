// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dependabot allowing allows testlib only 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    allow:
      - dependency-name: testlib
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot assignees empty assignees 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot assignees multiple assignees 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    assignees:
      - testUserOne
      - testUserTwo
"
`;

exports[`dependabot assignees no assignees 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot assignees one assignee 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    assignees:
      - testUserOne
"
`;

exports[`dependabot default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot groups applies-to 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    groups:
      group-one:
        applies-to: security-updates
        patterns:
          - "*"
"
`;

exports[`dependabot groups dependency-type 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    groups:
      group-one:
        dependency-type: development
        patterns:
          - "*"
"
`;

exports[`dependabot groups no groups 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot groups two groups 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    groups:
      group-one:
        patterns:
          - testlib-*
      group-two:
        patterns:
          - "*"
        exclude-patterns:
          - otherlib-*
"
`;

exports[`dependabot groups update-types 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    groups:
      group-one:
        update-types:
          - minor
          - patch
        patterns:
          - "*"
"
`;

exports[`dependabot ignoring ignore with a single version 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: testlib
        versions:
          - ">10.x"
      - dependency-name: projen
"
`;

exports[`dependabot ignoring ignore with ignoreProjen set to false 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
"
`;

exports[`dependabot ignoring ignore with multiple versions 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: testlib
        versions:
          - 10.x
          - 20.x
      - dependency-name: projen
"
`;

exports[`dependabot ignoring ignore with no version 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: testlib
      - dependency-name: projen
"
`;

exports[`dependabot ignoring ignores projen by default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot open-pull-requests-limit open pull requests limit not set 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot open-pull-requests-limit open pull requests limit set to 0 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    open-pull-requests-limit: 0
"
`;

exports[`dependabot open-pull-requests-limit open pull requests limit set to positive integer 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    open-pull-requests-limit: 5
"
`;

exports[`dependabot private registries 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
registries:
  npm-registry-npm-pkg-github-com:
    type: npm-registry
    url: https://npm.pkg.github.com
    token: \${{ secrets.TOKEN }}
    replaces-base: true
    username: test
    password: \${{ secrets.TOKEN }}
    key: abc123
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    registries:
      - npm-registry-npm-pkg-github-com
"
`;

exports[`dependabot reviewers empty reviewers 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot reviewers multiple reviewers 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    reviewers:
      - testUserOne
      - testUserTwo
"
`;

exports[`dependabot reviewers no reviewers 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;

exports[`dependabot reviewers one reviewer 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    reviewers:
      - testUserOne
"
`;

exports[`dependabot target branch develop as the target branch 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
    target-branch: develop
"
`;

exports[`dependabot target branch empty target branch 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

version: 2
updates:
  - package-ecosystem: npm
    versioning-strategy: lockfile-only
    directory: /
    schedule:
      interval: daily
    ignore:
      - dependency-name: projen
"
`;
